How to represent oneof with a schema/use oneof in .proto files?

When trying to generate a schema from a proto file, I get the following error:

```
/home/andre/hello-grpc/src/Schema.hs:21:1-42: Splicing declarations
    grpc "TheSchema" id "src/hello-grpc.proto"
  ======>
    type TheSchema =
        '[ 'DRecord "HelloRequest" '[ 'FieldDef "name" ('TPrimitive Text)],
           'DRecord "HelloReply" '[ 'FieldDef "message" ('TPrimitive Text)],
           'DRecord "Either" '[ 'FieldDef "inner" '[ 'TSchematic "HelloRequest",
                                                     'TSchematic "HelloReply"]]]
    type instance AnnotatedSchema Mu.Adapter.ProtoBuf.ProtoBufAnnotation TheSchema = '[ 'AnnField "HelloRequest" "name" ('Mu.Adapter.ProtoBuf.ProtoBufId 1 'True),
                                                                                        'AnnField "HelloReply" "message" ('Mu.Adapter.ProtoBuf.ProtoBufId 1 'True),
                                                                                        'AnnField "Either" "inner" ('Mu.Adapter.ProtoBuf.ProtoBufOneOfIds '[1,
                                                                                                                                                            2])]
    type Service =
        'Mu.Rpc.Package ('Just "hello_grpc") '[ 'Mu.Rpc.Service "Service" '[] '[ 'Mu.Rpc.Method "SayHelloStream" '[] '[ 'Mu.Rpc.ArgSingle 'Nothing '[] ('Mu.Rpc.SchemaRef TheSchema "HelloRequest")] ('Mu.Rpc.RetStream ('Mu.Rpc.SchemaRef TheSchema "HelloReply"))]]

/home/andre/hello-grpc/src/Schema.hs:21:1: error:
    • Expected kind ‘FieldTypeB * ghc-prim-0.5.3:GHC.Types.Symbol’,
        but ‘'[ 'TSchematic "HelloRequest",
                'TSchematic "HelloReply"]’ has kind ‘[FieldTypeB
                                                        builtin0 ghc-prim-0.5.3:GHC.Types.Symbol]’
    • In the second argument of ‘'FieldDef’, namely
        ‘'[ 'TSchematic "HelloRequest", 'TSchematic "HelloReply"]’
      In the second argument of ‘'DRecord’, namely
        ‘'[ 'FieldDef "inner" '[ 'TSchematic "HelloRequest",
                                 'TSchematic "HelloReply"]]’
      In the type ‘'[ 'DRecord "HelloRequest" '[ 'FieldDef "name" ('TPrimitive Text)],
                      'DRecord "HelloReply" '[ 'FieldDef "message" ('TPrimitive Text)],
                      'DRecord "Either" '[ 'FieldDef "inner" '[ 'TSchematic "HelloRequest",
                                                                'TSchematic "HelloReply"]]]’
   |
21 | grpc "TheSchema" id "src/hello-grpc.proto"
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Here is the proto file
```proto
syntax = "proto3";

import "google/protobuf/empty.proto";

package hello_grpc;

service Service {
  //rpc SayHello (HelloRequest) returns (HelloReply) {}
  rpc SayHelloStream (HelloRequest) returns (stream HelloReply) {}
}

message HelloRequest { string name = 1; }
message HelloReply { string message = 1; }

message Either {
  oneof inner {
    string name = 1;
    string message = 2;
  }
}
```

Looking at the definitions of the schema types this doesn't seem possible. Is this true?

--- a/adapter/protobuf/src/Mu/Quasi/ProtoBuf.hs
+++ b/adapter/protobuf/src/Mu/Quasi/ProtoBuf.hs
@@ -110,12 +110,12 @@ pbTypeDeclToType (P.DMessage name _ _ fields _) = do
       | not (all hasFieldNumber vs)
       = fail "nested oneof fields are not supported"
       | otherwise
-      = (,) <$> [t| 'FieldDef $(textToStrLit nm) $(typesToList <$> mapM pbOneOfFieldToType vs ) |]
+      = (,) <$> [t| 'FieldDef $(textToStrLit nm) ('TUnion $(typesToList <$> mapM pbOneOfFieldToType vs )) |]
             <*> [t| 'AnnField $(textToStrLit name) $(textToStrLit nm)
                        ('ProtoBufOneOfIds $(typesToList <$> mapM (intToLit . getFieldNumber) vs )) |]
 
     reportDefaultWarning :: [P.Option] -> Q ()
-    reportDefaultWarning opts = do
+    reportDefaultWarning opts =
       when (any (\(P.Option ident _) -> ident == ["default"]) opts)
            (reportError "mu-protobuf does not (yet) support default values")
 

--- a/core/schema/src/Mu/Schema/Class.hs
+++ b/core/schema/src/Mu/Schema/Class.hs
@@ -311,6 +311,10 @@ instance {-# OVERLAPS #-} GToSchemaFieldTypeUnion sch '[] U1 where
   toSchemaFieldTypeUnion U1 = error "this should never happen"
 instance {-# OVERLAPS #-} GFromSchemaFieldTypeUnion sch '[] U1 where
   fromSchemaFieldTypeUnion _ = U1
+instance {-# OVERLAPS #-} GToSchemaFieldTypeUnion sch '[] (M1 i t U1) where
+  toSchemaFieldTypeUnion (M1 U1) = error "this should never happen"
+instance {-# OVERLAPS #-} GFromSchemaFieldTypeUnion sch '[] (M1 i t U1) where
+  fromSchemaFieldTypeUnion _ = M1 U1
 instance {-# OVERLAPPABLE #-}
          TypeError ('Text "the type does not match the union")
          => GToSchemaFieldTypeUnion sch '[] f where
@@ -327,6 +331,15 @@ instance (GFromSchemaFieldTypeWrap sch t v)
          => GFromSchemaFieldTypeUnion sch '[t] v where
   fromSchemaFieldTypeUnion (Z x) = fromSchemaFieldTypeW x
   fromSchemaFieldTypeUnion (S _) = error "this should never happen"
+
+-- remove M1 from thing with more than one element
+instance {-# OVERLAPS #-} (GToSchemaFieldTypeUnion sch (a ': b ': rest) v)
+         => GToSchemaFieldTypeUnion sch (a ': b ': rest) (M1 i t v) where
+  toSchemaFieldTypeUnion (M1 x) = toSchemaFieldTypeUnion x
+instance {-# OVERLAPS #-} (GFromSchemaFieldTypeUnion sch (a ': b ': rest) v)
+         => GFromSchemaFieldTypeUnion sch (a ': b ': rest) (M1 i t v) where
+  fromSchemaFieldTypeUnion x = M1 (fromSchemaFieldTypeUnion x)
+
 instance (GToSchemaFieldTypeWrap sch t v, GToSchemaFieldTypeUnion sch ts vs)
          => GToSchemaFieldTypeUnion sch (t ': ts) (v :+: vs) where
   toSchemaFieldTypeUnion (L1 x) = Z (toSchemaFieldTypeW x)

